### PR TITLE
Implement `geom_equals` and binary predicates that depend only on it.

### LINKED ISSUE #832: [FEA]: Implement binary predicates that depend on `equals`
### Is this a new feature, an improvement, or a change to existing functionality?

New Feature

### How would you describe the priority of this feature request

Critical (currently preventing usage)

### Please provide a clear description of problem you would like to solve.

**Is your feature request related to a problem? Please describe.**
Some spatial predicates depend only on the outcome of point-in-polygon.

**Describe the solution you'd like**
I can work on these predicates in python while the `equals` and `intersects` basic predicates are being worked on.

Equality/Intersection/PIP table
Legend:
x: False/impossible
e: uses equality
i: uses intersection
p: uses pip

  | x(p, p) | x(p, l) | x(p, P) | x(l, p) | x(l, l) | x(l, P) | x(P, p) | x(P, l) | x(P, P)
-- | -- | -- | -- | -- | -- | -- | -- | -- | --
Equals | e | x | x | x | e | e | e | e | e
Disjoint | e | e | e | i | i | ip | ep | ep | ep
Touches | e | e | p | i | i | ip | ip | ip | ip
Contains | e | x | x | i | e | x | p | p | p
Covers | e | x | x | i | e | x | ip | ip | ip
Intersects | e | e | p | i | i | i | ip | ip | ip
Within | e | e | p | x | e | p | x | x | p
CoveredBy | e | e | p | x | e | p | x | x | p
Crosses | e | x | x | i | i | i | p | i | i
Overlaps | e | x | x | i | i | i | p | i | i

https://github.com/rapidsai/cuspatial/issues/832#tasklist-block-560d4b07-f265-48d6-be82-c736c9c583d4

### Describe any alternatives you have considered

_No response_

### Additional context

_No response_

### LINKED ISSUE #838: [FEA]: Binary predicates that depend only on equals
### Is this a new feature, an improvement, or a change to existing functionality?

New Feature

### How would you describe the priority of this feature request

Critical (currently preventing usage)

### Please provide a clear description of problem you would like to solve.

Many binary operations depend only on tests for equality. This issue is an epic for planning, tracking, and completing the binary operations depending only on equality.

Implement the binary operations that depend on equality. These operations include:
### Strict equality: `len(lhs) == len(rhs) && count_equals(lhs, rhs) == len(rhs)`
1. point equals point
3. linestring equals linestring
4. linestring equals polygon
7. polygon equals linestring
- What about closed vs. unclosed polygons?
8. polygon equals polygon
13. point contains point
15. point covers point
17. point intersects point
19. point within point
22. point coveredby point
25. point crosses point
26. point overlaps point

### Weak equality: `len(lhs) > len(rhs) && count_equals(lhs, rhs) == len(rhs)`
6. polygon equals point
14. linestring contains linestring
17. linestring covers linestring
18. point intersects linestring
20. point within linestring
21. linestring within linestring

### Strict inequality: `count_equals(lhs, rhs) == 0`
10. point disjoint point
11. point disjoint linestring
12. point disjoint polygon

Each of this binops depends on particular counting operations to occur as postprocessing. This issue also tracks those operations.

<b style="font-weight:normal;" id="docs-internal-guid-da5a398c-7fff-f004-66cb-6989e0b380d7"><h2 dir="ltr" style="line-height:1.38;margin-top:18pt;margin-bottom:6pt;"><span style="font-size:16pt;font-family:Arial;color:#000000;background-color:transparent;font-weight:400;font-style:normal;font-variant:normal;text-decoration:none;vertical-align:baseline;white-space:pre;white-space:pre-wrap;">Equality/Intersection/PIP table</span></h2><p dir="ltr" style="line-height:1.38;margin-top:0pt;margin-bottom:0pt;"><span style="font-size:11pt;font-family:Arial;color:#000000;background-color:transparent;font-weight:400;font-style:normal;font-variant:normal;text-decoration:none;vertical-align:baseline;white-space:pre;white-space:pre-wrap;">Legend:</span></p><p dir="ltr" style="line-height:1.38;margin-top:0pt;margin-bottom:0pt;"><span style="font-size:11pt;font-family:Arial;color:#000000;background-color:transparent;font-weight:400;font-style:normal;font-variant:normal;text-decoration:none;vertical-align:baseline;white-space:pre;white-space:pre-wrap;">x: False/impossible</span></p><p dir="ltr" style="line-height:1.38;margin-top:0pt;margin-bottom:0pt;"><span style="font-size:11pt;font-family:Arial;color:#000000;background-color:transparent;font-weight:400;font-style:normal;font-variant:normal;text-decoration:none;vertical-align:baseline;white-space:pre;white-space:pre-wrap;">e: uses equality</span></p><p dir="ltr" style="line-height:1.38;margin-top:0pt;margin-bottom:0pt;"><span style="font-size:11pt;font-family:Arial;color:#000000;background-color:transparent;font-weight:400;font-style:normal;font-variant:normal;text-decoration:none;vertical-align:baseline;white-space:pre;white-space:pre-wrap;">i: uses intersection</span></p><p dir="ltr" style="line-height:1.38;margin-top:0pt;margin-bottom:0pt;"><span style="font-size:11pt;font-family:Arial;color:#000000;background-color:transparent;font-weight:400;font-style:normal;font-variant:normal;text-decoration:none;vertical-align:baseline;white-space:pre;white-space:pre-wrap;">p: uses pip</span></p><div dir="ltr" style="margin-left:0pt;" align="left">

  | x(p, p) | x(p, l) | x(p, P) | x(l, p) | x(l, l) | x(l, P) | x(P, p) | x(P, l) | x(P, P)
-- | -- | -- | -- | -- | -- | -- | -- | -- | --
Equals | e | x | x | x | e | e | e | e | e
Disjoint | e | e | e | i | i | ip | ep | ep | ep
Touches | ei | ei | ip | ei | ei | ip | ip | ip | ip
Contains | e | x | x | i | e | x | p | p | p
Covers | e | x | x | i | e | x | ip | ip | ip
Intersects | e | i | p | i | i | i | ip | ip | ip
Within | e | e | p | x | e | p | x | x | p
CoveredBy | e | e | p | x | e | p | x | x | x
Crosses | e | x | x | i | i | i | x | ip | ip
Overlaps | e | x | x | i | i | i | p | ip | ip

</div></b>

### Describe any alternatives you have considered

_No response_

### Additional context

_No response_

### PR DESCRIPTION
**Pull request by thomcom**
Closes #838
Closes #832 
Depends on http://github.com/rapidsai/cudf/pull/12749

CI failure in this PR is caused by the above bug in CUDF.

## Description
This PR implements binary predicates that depend only on equality, which is implemented here using columnar comparison in python.

I'm playing with benchmarks of this feature now. On only Point geometries, we begin to outperform geopandas at 50k points, with 60x performance at 10m points.

## Checklist
- [x] I am familiar with the [Contributing Guidelines](https://github.com/rapidsai/cuspatial/blob/HEAD/CONTRIBUTING.md).
- [x] New or existing tests cover these changes.
- [x] The documentation is up to date with these changes.

### CODE REVIEW

File: python/cuspatial/cuspatial/core/geoseries.py [hunk 1]
```diff
@@ -209,9 +215,9 @@ def part_offset(self):
         def point_indices(self):
             # Return a cupy.ndarray containing the index values from the
             # LineString GeoSeries that each individual point is member of.
-            offsets = self.part_offset.take(self.geometry_offset)
+            offsets = cp.array(self.part_offset)
             sizes = offsets[1:] - offsets[:-1]
-            return cp.repeat(self._series.index, sizes)
+            return cp.repeat(self._meta.input_types.index, sizes)
```
Comment (thomcom) [inline python/cuspatial/cuspatial/core/geoseries.py:220]: I think this still needs to be updated, a regression relative to what has been merged lately.

File: python/cuspatial/cuspatial/core/geoseries.py [hunk 2]
```diff
@@ -874,14 +880,90 @@ def contains_properly(self, other, align=True):
...
+            [
+                Polygon([[0, 0], [1, 0], [1, 1], [0, 0]]),
+            ]
+        )
+        >>> polygon2 = cuspatial.GeoSeries(
+            [
+                Polygon([[0, 0], [1, 0], [1, 1], [0, 0]]),
+            ]
+        )
+        >>> print(polygon.equals(polygon2))
```
Comment (thomcom) [inline python/cuspatial/cuspatial/core/geoseries.py:928]: This needs to be renamed to `geom_equals`

**Comment by harrism** [comment]
The design of the binpreds eludes me. I really want to see it and the math behind each operation documented as discussed in #854

File: python/cuspatial/cuspatial/tests/binpreds/test_equals_only_binops.py [hunk 3]
```diff
@@ -0,0 +1,570 @@
...
+    points1 = cuspatial.from_geopandas(gpdpoints1)
+    points2 = cuspatial.from_geopandas(gpdpoints2)
+    got = points1.geom_equals(points2)
+    expected = gpdpoints1.geom_equals(gpdpoints2)
+    pd.testing.assert_series_equal(expected, got.to_pandas())
+
+
+def test_linestring_geom_equals_linestring():
+    gpdline1 = gpd.GeoSeries([LineString([(0, 0), (1, 1)])])
+    gpdline2 = gpd.GeoSeries([LineString([(0, 0), (1, 1)])])
```
Comment (isVoid) [inline python/cuspatial/cuspatial/tests/binpreds/test_equals_only_binops.py:63]: What happens if
```suggestion
    gpdline1 = gpd.GeoSeries([LineString([(0, 0), (1, 1)])])
    gpdline2 = gpd.GeoSeries([LineString([(1, 1), (0, 0)])])
```
?

File: python/cuspatial/cuspatial/tests/binpreds/test_equals_only_binops.py [hunk 4]
```diff
@@ -0,0 +1,570 @@
...
+    polygons = cuspatial.from_geopandas(gpdpolygons)
+    lines = cuspatial.from_geopandas(gpdlines)
+    got = polygons.geom_equals(lines)
+    expected = gpdpolygons.geom_equals(gpdlines)
+    pd.testing.assert_series_equal(expected, got.to_pandas())
+
+
+def test_polygon_geom_equals_polygon():
+    gpdpolygon1 = gpd.GeoSeries(Polygon([[0, 0], [1, 0], [1, 1], [0, 0]]))
+    gpdpolygon2 = gpd.GeoSeries(Polygon([[0, 0], [1, 0], [1, 1], [0, 0]]))
```
Comment (isVoid) [inline python/cuspatial/cuspatial/tests/binpreds/test_equals_only_binops.py:163]: Similarly,
```suggestion
    gpdpolygon1 = gpd.GeoSeries(Polygon([[0, 0], [1, 0], [1, 1], [0, 0]]))
    gpdpolygon2 = gpd.GeoSeries(Polygon([[0, 0], [1, 1], [1, 0],  [0, 0]]))
```

File: python/cuspatial/cuspatial/tests/binpreds/test_equals_only_binops.py [hunk 5]
```diff
@@ -0,0 +1,570 @@
...
+    gpdpolygons1 = gpd.GeoSeries([*polygon_generator(10000, 0)])
+    gpdpolygons2 = gpd.GeoSeries([*polygon_generator(10000, 0)])
+    polygons1 = cuspatial.from_geopandas(gpdpolygons1)
+    polygons2 = cuspatial.from_geopandas(gpdpolygons2)
+    got = polygons1.geom_equals(polygons2)
+    expected = gpdpolygons1.geom_equals(gpdpolygons2)
+    pd.testing.assert_series_equal(expected, got.to_pandas())
+
+
+def test_point_contains_point():
```
Comment (isVoid) [inline python/cuspatial/cuspatial/tests/binpreds/test_equals_only_binops.py:217]: All test below this points are testing `contains_properly` I suppose?

File: python/cuspatial/cuspatial/tests/binpreds/test_equals_only_binops.py [hunk 6]
```diff
@@ -0,0 +1,570 @@
...
+
+def test_10000_polygons_geom_equals_10000_polygons(polygon_generator):
+    gpdpolygons1 = gpd.GeoSeries([*polygon_generator(10000, 0)])
+    gpdpolygons2 = gpd.GeoSeries([*polygon_generator(10000, 0)])
+    polygons1 = cuspatial.from_geopandas(gpdpolygons1)
+    polygons2 = cuspatial.from_geopandas(gpdpolygons2)
+    got = polygons1.geom_equals(polygons2)
+    expected = gpdpolygons1.geom_equals(gpdpolygons2)
+    pd.testing.assert_series_equal(expected, got.to_pandas())
+
```
Comment (isVoid) [inline python/cuspatial/cuspatial/tests/binpreds/test_equals_only_binops.py:215]: Perhaps we should add multi-geometry test too?
```suggestion
def test_multipoint_geom_equals_point():
    gpdpoint1 = gpd.GeoSeries([MultiPoint([(0, 0), (1, 1)])])
    gpdpoint2 = gpd.GeoSeries([MultiPoint([(1, 1), (0, 0)])])
    point1 = cuspatial.from_geopandas(gpdpoint1)
    point2 = cuspatial.from_geopandas(gpdpoint2)
    got = point1.geom_equals(point2)
    expected = gpdpoint1.geom_equals(gpdpoint2)
    pd.testing.assert_series_equal(expected, got.to_pandas())
```
Similarly for multipolygons, multilinestrings.

**Review by isVoid — CHANGES_REQUESTED** [review]
I haven't looked into the implementation yet, but I have some question regarding test coverage.

File: python/cuspatial/cuspatial/tests/binpreds/test_equals_only_binops.py (hunk 5 above)
Comment (thomcom) [reply python/cuspatial/cuspatial/tests/binpreds/test_equals_only_binops.py]: No, these binary predicates depend only on `geom_equals` to return correct results. A point can contain a point in GeoPandas, and it appears to be a simple equality test.

File: python/cuspatial/cuspatial/tests/binpreds/test_equals_only_binops.py (hunk 6 above)
Comment (thomcom) [reply python/cuspatial/cuspatial/tests/binpreds/test_equals_only_binops.py]: Yes happy to.

File: python/cuspatial/cuspatial/tests/binpreds/test_equals_only_binops.py (hunk 4 above)
Comment (thomcom) [reply python/cuspatial/cuspatial/tests/binpreds/test_equals_only_binops.py]: This example fails! I'm working on a fix.

Comment (thomcom) [reply python/cuspatial/cuspatial/tests/binpreds/test_equals_only_binops.py]: I've discovered that it has to do with the closed format - polygons that don't share end points don't have the same points, because the end point is duplicated. I'm working on stripping off the extra endpoint now.

**Review by harrism — COMMENTED** [review]
I'm worried that unlike other features in cuSpatial, this one has very complex logic that is implemented only in Python, and not in our base C++. So supporting it in other languages, like Java, is not possible without rewriting a LOT of code.

It's also hard to reason about what and where computation is happening in CUDA.

File: python/cuspatial/cuspatial/core/binpreds/binpreds.py [hunk 7]
```diff
@@ -234,3 +266,229 @@ def preprocess(self, lhs, rhs):
...
+        ].reset_index(drop=True)
+        swap_rx = rhs_x[indices.last()].reset_index(drop=True) < rhs_x[
+            indices.first()
+        ].reset_index(drop=True)
+        swap_ry = rhs_y[indices.last()].reset_index(drop=True) < rhs_y[
+            indices.first()
+        ].reset_index(drop=True)
+        # Swap the first and last values of each linestring
+        (
+            lhs_x.iloc[indices.last()[swap_lx]],
```
[the remark this answers is not in the document]
Comment (thomcom) [reply python/cuspatial/cuspatial/core/binpreds/binpreds.py:347]: I was planning on putting them in a function. Checking if they are out of order and early terminating might be a good solution I'll look at.

File: python/cuspatial/cuspatial/tests/binpreds/test_equals_only_binops.py (hunk 4 above)
Comment (thomcom) [reply python/cuspatial/cuspatial/tests/binpreds/test_equals_only_binops.py]: This is fixed.

File: python/cuspatial/cuspatial/tests/binpreds/test_equals_only_binops.py (hunk 3 above)
Comment (thomcom) [reply python/cuspatial/cuspatial/tests/binpreds/test_equals_only_binops.py]: Because this isn't the final iteration of `geom_equals`, I'm going to settle for the test case that I have so far, which I think covers most usage. As I refactor the design and work through the intersection cases I will create exhaustive tests for all combinations and types.

File: python/cuspatial/cuspatial/tests/binpreds/test_equals_only_binops.py (hunk 6 above)
Comment (thomcom) [reply python/cuspatial/cuspatial/tests/binpreds/test_equals_only_binops.py]: For now I'd like to leave it with the functionality it has so far. I'll be refactoring the underlying architecture to handle every combination of types and orderings in the next couple of revisions.

File: python/cuspatial/cuspatial/core/binpreds/binpreds.py [hunk 8]
```diff
@@ -234,3 +266,229 @@ def preprocess(self, lhs, rhs):
...
+        return a & b
+
+    def _op(self, lhs, rhs):
+        indices = lhs.point_indices()
+        result = self._vertices_equals(lhs.xy, rhs.xy)
+        result_df = cudf.DataFrame(
+            {"idx": indices[: len(result)], "equals": result}
+        )
+        gb_idx = result_df.groupby("idx")
+        result = (gb_idx.sum().sort_index() == gb_idx.count().sort_index())[
```
[the remark this answers is not in the document]
Comment (thomcom) [reply python/cuspatial/cuspatial/core/binpreds/binpreds.py:475]: $a_i == b_i iff \forall j in len(a) a_j == b_j$

File: python/cuspatial/cuspatial/core/binpreds/binpreds.py [hunk 9]
```diff
@@ -43,15 +59,19 @@ def preprocess(self, lhs, rhs):
...
-    def postprocess(self, point_indices, point_result):
+    def postprocess(
+        self, point_indices: cudf.Series, point_result: cudf.Series
+    ) -> cudf.Series:
         """Postprocess the output data for the binary predicate. This method
         should be implemented by subclasses.
 
         Postprocess converts the raw results of the binary predicate into
-        the final result. This is where the discrete math rules are applied.
+        the final result. At this step the results for none, any, and all
```
[the remark this answers is not in the document]
Comment (thomcom) [reply python/cuspatial/cuspatial/core/binpreds/binpreds.py:73]: done

File: python/cuspatial/cuspatial/core/binpreds/binpreds.py (hunk 7 above)
[the remark this answers is not in the document]
Comment (thomcom) [reply python/cuspatial/cuspatial/core/binpreds/binpreds.py]: This is refactored nicely.

**Review by isVoid — COMMENTED** [review]
Nit picking more corner cases for `geom_equal` and want more tests for `overlaps` and `crossings`. But ok to punt for future if need to meet timing requirement.

File: python/cuspatial/cuspatial/core/binpreds/binpreds.py [hunk 10]
```diff
@@ -234,3 +267,233 @@ def preprocess(self, lhs, rhs):
...
+                return self._cancel_op(lhs, rhs, lengths_equal)
+        elif contains_only_polygons(lhs):
+            geoms_equal = self._offset_equals(
+                lhs.polygons.part_offset, rhs.polygons.part_offset
+            )
+            lengths_equal = self._offset_equals(
+                lhs[geoms_equal].polygons.ring_offset,
+                rhs[geoms_equal].polygons.ring_offset,
+            )
+            if lengths_equal.any():
```
[the remark this answers is not in the document]
Comment (thomcom) [reply python/cuspatial/cuspatial/core/binpreds/binpreds.py:440]: It is used on line 437 and 438 to gather only the geometries that are of equal length.

File: python/cuspatial/cuspatial/core/binpreds/binpreds.py [hunk 11]
```diff
@@ -234,3 +267,233 @@ def preprocess(self, lhs, rhs):
...
+                # No lengths are equal, so none can be equal.
+                return self._cancel_op(lhs, rhs, lengths_equal)
+        elif contains_only_linestrings(lhs):
+            lengths_equal = self._offset_equals(
+                lhs.lines.part_offset, rhs.lines.part_offset
+            )
+            if lengths_equal.any():
+                # Linestrings are equal if their sorted points
+                # are equal. This is unintuitive and perhaps
+                # incorrect, but it is the behavior of shapely.
```
[the remark this answers is not in the document]
Comment (thomcom) [reply python/cuspatial/cuspatial/core/binpreds/binpreds.py:424]: Thanks Michael, the functionality works as intended, I think this is a stale comment.

File: python/cuspatial/cuspatial/core/binpreds/binpreds.py [hunk 12]
```diff
@@ -234,3 +267,233 @@ def preprocess(self, lhs, rhs):
...
+        y = y.dropna()
+        new_offsets = self.lhs.polygons.ring_offset - cp.arange(
+            len(self.lhs.polygons.ring_offset)
+        )
+        xy = cudf.DataFrame({"x": x, "y": y}).interleave_columns()
+        return cuspatial.GeoSeries.from_multipoints_xy(
+            xy, new_offsets.astype("int32")
+        )
+
+    def _sort_polygons(self, lhs, rhs, initial):
```
[the remark this answers is not in the document]
Comment (thomcom) [reply python/cuspatial/cuspatial/core/binpreds/binpreds.py:381]: Ah, you are right. I can't handle this case with equals without iterating I think. `shapely.equals` written text says `If A is within B and B is within A, A and B are considered equal. The ordering of points can be different.` https://shapely.readthedocs.io/en/stable/reference/shapely.equals.html

It turns out that this case depends on using an intersection-enabled `.contains`, not `.contains_properly`. Thanks for noticing. I'll drop this case and the polygon cases which don't depend on `equals` after all.

[the remark this answers is not in the document]
Comment (thomcom) [reply python/cuspatial/cuspatial/core/binpreds/binpreds.py:381]: I could write this with a convolution easily but that too would require iteration I think, best to wait until `within` has intersection.

File: python/cuspatial/cuspatial/core/binpreds/binpreds.py [hunk 13]
```diff
@@ -2,18 +2,35 @@
...
 from cuspatial.utils.column_utils import (
     contains_only_linestrings,
     contains_only_multipoints,
+    contains_only_points,
     contains_only_polygons,
     has_same_geometry,
 )
 
 
+class PreprocessorOutput:
```
[the remark this answers is not in the document]
Comment (thomcom) [reply python/cuspatial/cuspatial/core/binpreds/binpreds.py:20]: Done.

File: python/cuspatial/cuspatial/core/binpreds/binpreds.py [hunk 14]
```diff
@@ -234,3 +267,233 @@ def preprocess(self, lhs, rhs):
...
+        )
+        (y.iloc[indices.last()[swap_y]], y.iloc[indices.first()[swap_y]],) = (
+            y.iloc[indices.first()[swap_y]],
+            y.iloc[indices.last()[swap_y]],
+        )
+        return cudf.Series(cp.column_stack((x.values, y.values)))
+
+    def _sort_linestrings(self, lhs, rhs, initial):
+        """Swap first and last values of each linestring to ensure that
+        the first point is the lowest value. This is necessary to ensure
```
[the remark this answers is not in the document]
Comment (thomcom) [reply python/cuspatial/cuspatial/core/binpreds/binpreds.py:351]: I can't seem to reply to all of the comments so I'll address them here.

1. It is possible with a pile of conditional logic to compute the group sizes via their offsets buffers, but using `object.point_indices()` and `groupby` allows me to avoid implementing those code paths. It is a "time of development vs. time of execution" problem, and the overall cost to a python user is not significant I think.

2. Not sure what you mean.

3. Summing a bool column with 255 values in `cudf` returns an `int64`.

4. It is lowest in the context of the endpoints, the first point needs to be lexicographically before the last point in order for comparison of all the points via sorting to work. I updated the docs.

File: python/cuspatial/cuspatial/core/binpreds/binpreds.py [hunk 15]
```diff
@@ -234,3 +267,233 @@ def preprocess(self, lhs, rhs):
...
+            x.iloc[indices.first()[swap_x]],
+            x.iloc[indices.last()[swap_x]],
+        )
+        (y.iloc[indices.last()[swap_y]], y.iloc[indices.first()[swap_y]],) = (
+            y.iloc[indices.first()[swap_y]],
+            y.iloc[indices.last()[swap_y]],
+        )
+        return cudf.Series(cp.column_stack((x.values, y.values)))
+
+    def _sort_linestrings(self, lhs, rhs, initial):
```
[the remark this answers is not in the document]
Comment (thomcom) [reply python/cuspatial/cuspatial/core/binpreds/binpreds.py]: I renamed it.

File: python/cuspatial/cuspatial/core/binpreds/binpreds.py [hunk 16]
```diff
@@ -234,3 +267,233 @@ def preprocess(self, lhs, rhs):
...
+            lhs_polygons_as_multipoints.multipoints,
+            rhs_polygons_as_multipoints.multipoints,
+            initial,
+        )
+
+    def preprocess(self, lhs, rhs):
+        # Compare types
+        type_compare = lhs.dtype == rhs.dtype
+        # Any unmatched type is not equal
+        if (type_compare == False).all():  # noqa: E712
```
[the remark this answers is not in the document]
Comment (thomcom) [reply python/cuspatial/cuspatial/core/binpreds/binpreds.py:376]: Because `lhs.dtype` returns a `Series` of dtypes, one for each row in the `GeoSeries`.

File: python/cuspatial/cuspatial/core/binpreds/binpreds.py (hunk 14 above)
[the remark this answers is not in the document]
Comment (isVoid) [reply python/cuspatial/cuspatial/core/binpreds/binpreds.py]: Can you move replies to: https://github.com/rapidsai/cuspatial/pull/926#discussion_r1134753854 ?

File: python/cuspatial/cuspatial/core/binpreds/binpreds.py (hunk 16 above)
[the remark this answers is not in the document]
Comment (isVoid) [reply python/cuspatial/cuspatial/core/binpreds/binpreds.py:376]: I've never seen a `dtype` property that returns a series instead of a scalar... This is new to me. Would you mind point me to where this property is documented?

File: python/cuspatial/cuspatial/tests/binpreds/test_equals_only_binpreds.py [hunk 17]
```diff
@@ -548,21 +482,34 @@ def test_3_linestrings_geom_equals_3_linestrings_longer():
...
+    )
+    linestring1 = cuspatial.from_geopandas(gpdlinestring1)
+    linestring2 = cuspatial.from_geopandas(gpdlinestring2)
+    got = linestring1.geom_equals(linestring2)
+    expected = gpdlinestring1.geom_equals(gpdlinestring2)
+    pd.testing.assert_series_equal(expected, got.to_pandas())
+
+
+"""These tests will be reintroduced when we have `.contains` implemented
+def test_pair_polygons_different_ordering():
```
Comment (isVoid) [inline python/cuspatial/cuspatial/tests/binpreds/test_equals_only_binpreds.py:504]: Use `pytest.skip(reason =...)`, not comment out.

File: python/cuspatial/cuspatial/core/binpreds/binpreds.py (hunk 14 above)
[the remark this answers is not in the document]
Comment (isVoid) [reply python/cuspatial/cuspatial/core/binpreds/binpreds.py]: And to your reply on this comment:

> It is lowest in the context of the endpoints, the first point needs to be lexicographically before the last point in order for comparison of all the points via sorting to work. I updated the docs.

Can you elaborate? How does the endpoints lexicographic order affects the equality of two linestrings?

**Review by isVoid — COMMENTED** [review]
I have a few more questions. It's definitely getting closer.

File: python/cuspatial/cuspatial/core/binpreds/binpreds.py (hunk 16 above)
[the remark this answers is not in the document]
Comment (thomcom) [reply python/cuspatial/cuspatial/core/binpreds/binpreds.py:376]: `GeoSeries.py:103`, it just returns the `self._column._meta_input_types`.

File: python/cuspatial/cuspatial/tests/binpreds/test_equals_only_binpreds.py (hunk 17 above)
Comment (thomcom) [reply python/cuspatial/cuspatial/tests/binpreds/test_equals_only_binpreds.py:504]: Done.

File: python/cuspatial/cuspatial/core/binpreds/binpreds.py [hunk 18]
```diff
@@ -185,18 +214,21 @@ def postprocess(self, point_indices, point_result):
...
-            # process for completed linestrings, polygons, and multipoints.
-            # Not necessary for points.
+            # Compute the set of results for each point-in-polygon predicate.
+            # Group them by the original index, and sum the results. If the
+            # sum of points in the rhs feature is equal to the number of
+            # points found in the polygon, then the polygon contains the
+            # feature.
             df_result = (
                 result.groupby("idx").sum().sort_index()
                 == result.groupby("idx").count().sort_index()
```
[the remark this answers is not in the document]
Comment (thomcom) [reply python/cuspatial/cuspatial/core/binpreds/binpreds.py:237]: In the preprocess step, https://github.com/rapidsai/cuspatial/blob/b7d4d1724cd115defdf9aa14658fd6eaa3b130cd/python/cuspatial/cuspatial/core/binpreds/binpreds.py#L172-L197, depending on rhs type the point values are extracted and flattened into a `GeoSeries` of points. At that time `point_indices` is also recorded. In order to not use point_indices here that logic would have to be reproduced in the `postprocess` as well.

File: python/cuspatial/cuspatial/core/geoseries.py [hunk 19]
```diff
@@ -96,6 +100,10 @@ def __init__(
             column, index, dtype=dtype, name=name, nan_as_null=nan_as_null
         )
 
+    @property
+    def dtype(self):
```
Comment (isVoid) [inline python/cuspatial/cuspatial/core/geoseries.py:104]: I don't think calling this `dtype` is appropriate. series.dtype returns a numpy type, geopandas returns `GeometryDtype`. They are all scalars.

File: python/cuspatial/cuspatial/tests/binpreds/test_equals_only_binpreds.py (hunk 17 above)
Comment (isVoid) [reply python/cuspatial/cuspatial/tests/binpreds/test_equals_only_binpreds.py:504]: It's not shown as done here. Have you pushed it?

**Review by isVoid — COMMENTED** [review]
Last couple questions.

File: python/cuspatial/cuspatial/core/binpreds/binpreds.py (hunk 14 above)
[the remark this answers is not in the document]
Comment (thomcom) [reply python/cuspatial/cuspatial/core/binpreds/binpreds.py]: It doesn't! Thanks for calling out another mistake. I had built some test cases around strings of length 2 and ended up believing the data had to do with the start and end of the strings. What really happens is that a linestring is equal to the opposite linestring, which works with sort, but doesn't work with linestrings constructed of the same point in different orders.

The updated code now reverses each linestring within its "bucket" along the GeoArrow x/y buffer, then does a forward and a reverse comparison and `or`s them.

File: python/cuspatial/cuspatial/tests/binpreds/test_equals_only_binpreds.py (hunk 17 above)
Comment (thomcom) [reply python/cuspatial/cuspatial/tests/binpreds/test_equals_only_binpreds.py]: Hmm I thought so.

File: python/cuspatial/cuspatial/core/geoseries.py (hunk 19 above)
Comment (thomcom) [reply python/cuspatial/cuspatial/core/geoseries.py:109]: Ah that's a good point.

Comment (thomcom) [reply python/cuspatial/cuspatial/core/geoseries.py]: Done

**Comment by thomcom** [comment]
Hope this is coming along to your liking @isVoid. As for more `overlaps` and `crosses`, point->point is the only geometry type that is supported yet, so there's only tests for that.
There's a passing test, a failing test, and then 10k randomly sampled coords tested. I've added a test that validates successful comparisons at index 0, 1, and 2 on separate calls as well.

File: python/cuspatial/cuspatial/tests/binpreds/test_equals_only_binpreds.py [hunk 20]
```diff
@@ -340,6 +357,24 @@ def test_point_not_overlaps_point():
...
+        [Point(3, 3), Point(3, 3), Point(2, 2)],
+    ],
+)
+def test_three_points_overlaps_three_points(points):
+    gpdpoints1 = gpd.GeoSeries(points)
+    gpdpoints2 = gpd.GeoSeries([Point(0, 0), Point(1, 1), Point(2, 2)])
+    points1 = cuspatial.from_geopandas(gpdpoints1)
+    points2 = cuspatial.from_geopandas(gpdpoints2)
+    got = points1.overlaps(points2)
+    expected = gpdpoints1.crosses(gpdpoints2)
```
Comment (isVoid) [inline python/cuspatial/cuspatial/tests/binpreds/test_equals_only_binpreds.py:374]:
```suggestion
    expected = gpdpoints1.overlaps(gpdpoints2)
```

File: python/cuspatial/cuspatial/core/binpreds/binpreds.py [hunk 21]
```diff
@@ -306,63 +305,108 @@ def _offset_equals(self, lhs, rhs):
...
+        "points": an index for each point in `coords`.
+        "xy_key": an index that maintains x/y ordering.
+        "xy": the x/y coordinates in `coords`.
+
+        The dataframe is sorted according to keys passed in by the caller.
+        For sorting multipoints, the keys in order are "object_key", "xy",
+        "xy_key". This sorts the points in each multipoint into the same
+        bin defined by "object_key", then sorts the points in each bin by
+        x/y coordinates, and finally sorts the points in each bin by x/y
+        ordering.
```
Comment (isVoid) [inline python/cuspatial/cuspatial/core/binpreds/binpreds.py:324]: > x/y ordering

Just checking, I assume you mean x coordinate should precede y coordinate?

**Review by isVoid — APPROVED** [review]
Thanks. There's a typo in the test and 1 question for doc. Approving.

File: python/cuspatial/cuspatial/core/binpreds/binpreds.py (hunk 21 above)
Comment (thomcom) [reply python/cuspatial/cuspatial/core/binpreds/binpreds.py:590]: Yes, that's the purpose of the `xy_key`.
